torch/testing/_internal/common_device_type.py

```diff
@@ -788,6 +788,72 @@ def test_wrapper(*args, **kwargs):
                     print("Failed to instantiate {0} for op {1}!".format(test_name, op.name))
                     raise ex
 
+# Decorator that generates two set of tests with layout=torch.strided and layout=specified_layout.
+# All keyword arguments are used only for the specified layout, so they have no effect on the original test.
+class enableLayout(_TestParametrizer):
+    def __init__(self, layout: torch.layout, *,
+                 dtypes: Optional[Sequence[torch.dtype]] = None,
+                 dtypes_cpu: Optional[Sequence[torch.dtype]] = None,
+                 dtypes_cuda: Optional[Sequence[torch.dtype]] = None,
+                 only_cuda: bool = False,
+                 skip_cuda_if_no_cusparse: bool = True):
+        if not isinstance(layout, torch.layout):
+            raise RuntimeError("Expected to get torch.layout as layout, but got ", layout)
+        self.layout = layout
+        self.dtypes = set(dtypes) if dtypes is not None else None
+        self.dtypes_cpu = set(dtypes_cpu) if dtypes_cpu is not None else None
+        self.dtypes_cuda = set(dtypes_cuda) if dtypes_cuda is not None else None
+        self.only_cuda = only_cuda
```

Comment:
What is the use case for not doing the test on CPU?

Reply:
When the tested function is implemented only for cuda and not for cpu. When I have the opposite situation I'd add `only_cpu`.

Reply:
Excuse my ignorance here, but do these cases actually exist? Do we have operators that only work on CUDA? Or do you mean a test that specifically deals with some CUDA behavior?

Regardless, can't we use the [`@onlyCPU` and `@onlyCUDA` decorators](https://github.com/pytorch/pytorch/blob/9b908ab0d0a947d89ac3137f8c4a05a87c35f568/torch/testing/_internal/common_device_type.py#L1076-L1081) to get what we need?

Reply:
During the development, it's rarely the case to implement the CUDA and CPU support simultaneously. For example in the current stack of PRs, first CUDA support is added, then CPU. So the operator can work only for CUDA, or only for CPU, or only for specific dtypes.
`onlyCUDA` decorator is used here if `only_cuda=True`: https://github.com/pytorch/pytorch/blob/d9de883b87ea39054bb99986775fc76b6fc54ece/torch/testing/_internal/common_device_type.py#L843-L844
dtype/device coverage of sparse_csr layout is very different from strided layout, that's why we need to keep these granular skips separate for these layouts. Instead of adding `@onlyCUDAIfSparseCSR` I added the flag here.

Reply:
> `onlyCUDA` decorator is used here if `only_cuda=True`

Either I still don't fully understand its purpose, but isn't 

```python
@enableSparseCSR(..., only_cuda=True)
def test_foo(self, device, dtype, layout):
    pass
```

equivalent to

```python
@onlyCUDA
@enableSparseCSR(...)
def test_foo(self, device, dtype, layout):
    pass
```

? If yes, I'm against having the `only_cuda` kwarg, because this functionality already exists in an convenient way.

> type/device coverage of sparse_csr layout is very different from strided layout, that's why we need to keep these granular skips separate for these layouts.

We have the `dtypes_cpu` and `dtypes_cuda` kwargs and I'm all for adding them. I currently don't see the connection to `only_cuda` though.

Reply:
Ok, I think I got it now. The use case is that you have an existing operator that already has CPU / CUDA support for strided tensors and you want to add CPU **or** CUDA sparse support for this operator, correct? If yes, sorry for the confusion.

---

Would it make sense to make this a little more general similar to the logic in `instantiate_device_type_tests`? Instead of having the `only_cuda` kwarg, have a `except_for` and `only_for` pair and use every device by default?